File: problem4/main.hs
revN_ n x =
    if n == 0 then x else
    revN_ (n `div` base) ((x * base) + (n `mod` base))
    where
        base = 10

revN n = revN_ n 0

isPalindrome n = revN n - n == 0

main =
    print $ foldl max 0 $
        filter isPalindrome [x * y | x <- [100..1000], y <- [100..1000]]

File: problem4/main.c
#include <stdio.h>
#include <stdbool.h>

#define BASE 10

bool is_palindrome(size_t n) {
    size_t number = n;

    // Construct reverse of n
    size_t rev_n = 0;
    while (number > 0) {
        size_t last_digit = number % BASE;
        rev_n = rev_n * BASE + last_digit;
        number /= BASE;
    }

    // if n is palindrome, difference of
    // its palindrome and itself is 0
    return n - rev_n == 0;
}

size_t largest() {
    size_t max = 0;
    for (size_t i = 999; i >= 100; i--) {
        for (size_t j = 999; j >= 100; j--) {
            size_t product = i * j;
            // printf("i: %d, p: %d, product: %d\n", i, j, product);
            if (is_palindrome(product))
                max = max < product ? product : max;
        }
    }

    return max;
}

int main() {

    printf("%d\n", largest());
    // 906609
    return 0;
}